### PR TITLE
Introducing Permission enum

### PR DESCRIPTION
**Pull request by joesus**
Introduces a permissions enum with known cases.

## Checklist

- [x] I've read the [Contributing Guidelines](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md)
- [x] I've completed the [Contributor License Agreement](https://developers.facebook.com/opensource/cla)
- [x] I've ensured that all existing tests pass and added tests (when/where necessary)
- [x] I've ensured that my code lints properly: (`swiftlint && swiftlint autocorrect --format`)
- [x] I've updated the documentation (when/where necessary) and [Changelog](CHANGELOG.md) (when/where necessary)
- [x] I've added the proper label to this pull request (e.g. `bug` for bug fixes)

## Pull Request Details

Introduces a Permission enum to use as the canonical type for representing permissions and declined permissions.

## Test Plan

Test Plan: Unit tests should all pass. Code coverage should remain at 100%

### CODE REVIEW

File: Sources/Core/Permission.swift [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  case userHometown
+  case userLikes
+  case userLink
+  case userLocation
+  case userMobilePhone
+  case userPhotos
+  case userPosts
+  case userTaggedPlaces
+  case userVideos
+  case unknown(value: String)
```
Comment (robtimp) [inline Sources/Core/Permission.swift]: Nit: wouldn't `other` be better than `unknown`?

Comment (joesus) [reply Sources/Core/Permission.swift]: I sort of think these are different. other is sort of like default so I wanted to avoid it. I see what you mean though. At some point there will be string values coming from the server that we need to turn into these cases. If we can't create a known case from the string then it's not a known case (as opposed to an `other` known case). <- sorry for the book, just wanted to make sure I actually thought about this enough.

File: Sources/Core/Permission.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import Foundation
+
+// TODO: Introduce a type to JSON encode permissions This object will be responsible for translating between JSON with snake_case keys and our canonical camelCased cases
+
+enum Permission: Hashable {
+  case `default`
```
Comment (robtimp) [inline Sources/Core/Permission.swift]: There's quite a few others: `publish_videos`, `pages_messaging`, all the various ones for Instagram and business permissions. Should we consider namespacing some of them (e.g. `Permission.Instagram.manageComments`)?

https://developers.facebook.com/docs/facebook-login/permissions#reference-default

File: Sources/Core/Permission.swift [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  // swiftlint:disable identifier_name
+  case id
```
Comment (robtimp) [inline Sources/Core/Permission.swift]: `id`, `firstName`, `lastName`, `middleName`, `name`, `nameFormat`, `picture` and `shortName` are all fields rather than permissions

File: Sources/Core/Permission.swift (hunk 2 above)
Comment (joesus) [reply Sources/Core/Permission.swift]: Yes and no. I had a few thoughts. 

1. I was trying to introduce the minimum at this point based on the yagni principle.
2. I have some vague idea that we'll be able to use framework name-spacing to accomplish this once we know where different permissions belong but I don't have this 100% fleshed out yet.

TLDR; I wanted to defer the decision of where to put the 'non-essential' permissions for as long as possible. Unless they all for sure belong in CoreKit in which case I'm happy to add them. :)

File: Sources/Core/Permission.swift (hunk 3 above)
Comment (joesus) [reply Sources/Core/Permission.swift]: I see. Will remove. Thanks!
